### DISCUSSION TITLE
OpenThread CLI - CoAPS Example

### DISCUSSION THREAD

**Discussion by janhard** [openthread/openthread#8462] [Q&A] [answered]
Hello! I have two NORDIC nRF52840 DKs that send sensor readings from one board to another over CoAP. I have tried to do the same using CoAP Secure, but otCoapSecureSendRequest() returns the error code 13 (OT_ERROR_INVALID_STATE) even though otCoapSecureConnect() does not return an error (OT_ERROR_NONE) and both otCoapSecureIsConnected as well as otCoapSecureIsConnectionActive() return TRUE.
However, since I couldn't figure out the cause for the unsuccessful SendRequest(), I flashed both my boards with the OpenThread CLI example (using nRF Connect SDK v2.0.0 and nRF Connect SDK v2.2.0-rc1) and followed the instructions in "README_COAPS.md". Even though I am following the instructions, a DTLS connection cannot be established ('coaps connect' returns the error message "Error 7: InvalidArgs" -> see attached screenshot). 

I appreciate any hints/help you may be able to provide for resolving my issue. Thank you!

![Screenshot 2022-11-30 180605](https://user-images.githubusercontent.com/95551723/204934425-b5fe9c8c-34dd-4552-a42c-cbc6652ec372.png)

**Comment by jwhui** [comment, marked as answer]
Are you able to reproduce the same behavior when using the Nordic nRF52840 example: https://github.com/openthread/ot-nrf528xx/blob/main/src/nrf52840 ?

**Reply by janhard** [reply]
Hello, and thank you very much for your response. I was able to flash that example, but it does not come with the coap(s) commands; hence I was unable to produce that behavior.

**Reply by jwhui** [reply]
Enable build features by appending CMake options to the end of the build command.
```
$ ./script/build nrf52840 USB_trans -DOT_BOOTLOADER=USB -DOT_COAP=ON -DOT_COAPS=ON
```

**Reply by janhard** [reply]
It took me a little longer to flash the devices this time. Please excuse the delayed response. I was able to reproduce the same behavior using the example you linked in your first response: coaps connect returns 'Error 7: InvalidArgs'. (please see attached screenshot)

![image](https://user-images.githubusercontent.com/95551723/205518665-6420188d-738d-4115-85e6-7b4b303afddb.png)

**Reply by janhard** [reply]
Never mind, I am unable to reproduce the same behavior. Using the mentioned example works, I used 'coaps set psk <psk> <pskid>' instead of 'coaps psk <psk> <pskid>':

Screenshot 1:
![image](https://user-images.githubusercontent.com/95551723/205551715-d0135b37-b759-47c4-aac4-5cc6ebb293c2.png)

In the newer example, however, the same command 'coaps psk <psk> <pskid>' returns "Error 35: InvalidCommand". I have also tried to use the certificate, but a connection is not being established successfully. 

Screenshot 2:
![image](https://user-images.githubusercontent.com/95551723/205553429-3f53f3cf-cc48-4b22-86ff-f3a002d11425.png)

Are there any examples of how to use the coaps_secure API? I tried to figure it out using the OpenThread CLI - CoAPS Example but was unsuccessful. For me, the description provided in the documentation (e.g., https://infocenter.nordicsemi.com/index.jsp?topic=%2Fsdk_tz_v3.2.0%2Fgroup__api-coap-secure.html) is not sufficient. I would be very grateful for any form of support.

**Reply by jwhui** [reply]
When using Zephyr, you need to prefix all OT CLI commands with `ot`.

The OT CLI implementation uses the `otCoapSecure` APIs. So I would just use OT CLI as an example for how to use the `otCoapSecure` APIs.
